### ISSUE TITLE
input blur event problems (Chrome)

### ISSUE THREAD

**Issue by yuretz** [WebReflection/lighterhtml#57]
In some cases our lighterhtml app is throwing the following error:
```
router.service.ts:512 Uncaught DOMException: Failed to execute 'removeChild' on 'Node': The node to be removed is no longer a child of this node. Perhaps it was moved in a 'blur' event handler?
    at remove (webpack:///./node_modules/domdiff/esm/utils.js?:90:12)
    at domdiff (webpack:///./node_modules/domdiff/esm/index.js?:161:61)
    at Object.anyContent [as fn] (webpack:///./node_modules/lighterhtml/esm/tagger.js?:240:81)
    at eval (webpack:///./node_modules/domtagger/esm/index.js?:89:20)
    at eval (webpack:///./node_modules/domtagger/esm/index.js?:116:21)
    at unroll (webpack:///./node_modules/lighterhtml/esm/index.js?:156:11)
    at update (webpack:///./node_modules/lighterhtml/esm/index.js?:125:20)
    at render (webpack:///./node_modules/lighterhtml/esm/index.js?:31:24)
    at update (webpack:///./src/app.ts?:88:74)
    at Object.dispatch (webpack:///./node_modules/redux/es/redux.js?:229:7)
```
I wasn't yet able to reproduce it outside our app, so I don't have any jsfiddle/codepen ready (though I'm still trying), but here is a description of what's going on and what I figured out so far. 

The issue seem to be caused by a text input element with an `onblur` handler that dispatches an action and causes lighterhtml `render()` to be called. The target element stays the same after re-render. 
After that, if I navigate to a different route (client-side) in our app (causing yet another `render()`, now changing the contents completely), then _sometimes_ I would see an error above. Removing `onblur` event or preventing it from causing a render makes the issue go away.  This issue seems to only happen in Chrome (tested in Chrome 76 on Windows 10), never seen anything like this in Firefox.

While investigating this issue, I found [this old answer on StackOverflow](https://stackoverflow.com/a/22934552), and it looks kind of similar, so I thought I would mention it here, maybe it could help solving this too.

**Comment by WebReflection** [comment]
do you have ternary operators in the mix?

**Comment by yuretz** [comment]
Yes, part of the page being rendered by the target route that provokes the exception looks something like:
```javascript
    (data) => html`
        <section class="ml-6 pt-6">
            <p>${data.section && data.section.description ? subcomponent(model.section.description) : ''}</p>
        </section>
    `;
```
Is it bad to do it this way?

**Comment by yuretz** [comment]
Another weird thing is that I cannot provoke it on my home laptop, though the setup should be identical... 🤔

**Comment by WebReflection** [comment]
what does `subcomponent(data.section.description)` returns?

**Comment by yuretz** [comment]
In our actual app it returns markdown converted to a DOM fragment. But it probably doesn't matter because I can still reproduce it (on my work machine) if I completely remove ternary expression like this: 
```javascript
    (data) => html`
        <section class="ml-6 pt-6">
            <p>foo</p>
        </section>
    `;
```
I still think that it's closely related to input element focus/blur handling because what I need to do in Chrome in order to reproduce it every time is:
1. Click on the misbehaving input and type something.
2. Without clicking anywhere else, click on the browser address bar. `blur` event and then re-render is triggered.
3. Manually type in a different route address and hit 'Enter'.
4. Boom! Exception is thrown while re-rendering everything

**Comment by WebReflection** [comment]
so you are confident about how to reproduce but you cannot provide a minimal example that fails? this is making it harder for me to understand what's going on.

> In our actual app it returns markdown converted to a DOM fragment.

Very bad, as the DOM fragment loses its children as son as injected, so that you trash the layout every single render, which is not the paradigm promoted by hyper/lighterhtml

Why don't you use `html` instead to return the subcomponent, so that the returned node is better understood by the stack?

```js
const subcomponent = data => html`something with ${data}`;
```

**Comment by yuretz** [comment]
Well, regarding the markdown part, we do it partially through lighterhtml. Roughly like this:
```javascript
const raw = (content, tag = html) => {
    const result = [content];
    result.raw = [content];
    return tag(result);
};

const subcomponent = (markdownString) => html`${raw(markdown2html(markdownString)}`;
```
`markdown2html` is not currently memoized, but it might make sense to do it when we get to production. Does it still look bad?

Regarding standalone minimal example, I totally understand and I am working on. 

It just takes more time, because it involves client-side routing (not easy in codepen/jsfiddle) and I also need to cut away unnecessary app pieces, and it's also not 100% clear what exactly triggers it, and it's also in Typescript, which you dislike, so I'm converting it to Javascript, and I also have some other stuff on my todo list. So things are a bit more complicated than usually, but I'll manage it sooner or later. 

And thanks again for your nice libraries, time and help you provide!

**Comment by WebReflection** [comment]
> markdown2html is not currently memoized, but it might make sense to do it when we get to production. Does it still look bad?

If you change it for prod it doesn't, but I suggest you doing that, otherwise it's still layout trashing per each render.

```js
let clearContents = 0;
const contents = new Map;
const raw = (content, tag = html) => {
    let result = contents.get(content);
    if (!result) {
      result = Object.freeze([content]);
      Object.defineProperty(result, 'raw', {value: result});
      contents.set(content, result);
      // used to auto cleanup the map to avoid leaks
      // useful only if content is heavily dynamic
      clearTimeout(clearContents);
      clearContents = setTimeout(() => contents.clear(), 60000);
    }
    return tag(result);
};
```

With above code you pass a better expected template-tag array, passing all the checks needed to know if the browser has standard templates or not (although this might lead to false positives 'cause [the check is much more complex than that](https://github.com/ungap/template-literal/blob/master/esm/index.js)).

> and it's also in Typescript, which you dislike

I don't just dislike it, TypeScript also produces [broken template literals tag arrays](https://github.com/ungap/template-literal/blob/master/esm/index.js#L22) so actually you don't need to meomize anything in there since TS is expected to inevitably fallback to the slow path, hence you can avoid even creating the `raw` bit in your example.

```js
// lighterhtml takes care of meomizing things for you as fallback
const raw = (content, tag = html) => tag([content]);
```

With TS that's all you need here.

> So things are a bit more complicated than usually, but I'll manage it sooner or later.

Looking forward to see an example, feel free to create a temporary repository that reproduces the issue with pure JS

**Comment by WebReflection** [comment]
P.S. actually, the template-literal fallback still trust unique Array, so this might be the best thing for your TS use case:

```js
const contents = new Map;
const raw = (content, tag = html) => {
    let result = contents.get(content);
    if (!result) {
      result = [content];
      result.raw = result;
      contents.set(content, result);
    }
    return tag(result);
};
```

In this way you create a TS broken like Array, but at least it's unique with same content, each time.

**Comment by WebReflection** [comment]
P.S.2 ... oh gosh, never mind, with TS this is still the best way to go:
```js
const raw = (content, tag = html) => tag([content]);
```

Apologies for my confusion, this template-literal inconsistency with TS and jurassic browsers is a mess

**Comment by yuretz** [comment]
I did it! You can take a look at a [minimal example here](https://gist.githack.com/yuretz/d721a0ff8ef5083d7d3d4ac2aa74fc42/raw/85cfce67a2474c701cc2b2b1ef7294b17943aba5/hyperblur.html). Works every time in my Chrome.

**Comment by yuretz** [comment]
Regarding TypeScript, we are using [@babel/preset-typescript](https://babeljs.io/docs/en/babel-preset-typescript) to transpile our TS code, and this means type declarations are just stripped and what's left is just JavaScript which then should be transpiled by babel as normal, resulting in [this flag](https://github.com/ungap/template-literal/blob/master/esm/index.js#L12) being false in our app at runtime.

**Comment by WebReflection** [comment]
OK, this seems to be a Chrome specific bug, where the element gets the focus again once you change the hash (popstate).

If you add `if (document.activeElement) document.activeElement.blur();` inside the `popstate` handler, everything is fine.

The issue, however, is that if the input is focused, the moment you `update(true)` with that ternary in place, the blur event fires synchronously **while** the element is being removed, and the `update(false)` call within the blur messes up the tree.

As it is for React hooks, ternary operator is not really a good pattern for the stack, but this concurrent change through an inconsistent state might be hard to solve, as Chrome revives the focus on an element that doesn't exist anymore, while it's blurring.

It might be a Chrome bug though, but I wonder what's the actual real-world scenario you have that needs that ternary, with such logic in place.

Since I don't want to handle `blur` as special event with issues, it'd be awesome to understand how rare, or easy, it's to have such error, as for today nobody had issues in general, even with the ternary.

TL;DR two concurrent states change on a focused element that might disappear will throw inevitably on Chrome, not sure there's really much I can do

**Comment by WebReflection** [comment]
P.S. the error itself states:

> The node to be removed is no longer a child of this node. Perhaps it was moved in a 'blur' event handler?

which is self explanatory of this issue. That's a DOM violation for Chrome, this library might not be able to fix it in a trivial way.

**Comment by yuretz** [comment]
Just to be clear, in our real world app there is no ternary expression at the top level. Instead the top level selects a component function to call based on the data extracted from the route part of address. The ternary expression in one of the components of our app, that I mentioned earlier, doesn't seem to play any significant role for this bug, since I can still reproduce it if I completely remove the ternary. Usage of ternary expression in the example was just my clumsy attempt to make it look simpler and remove unnecessary details from the demo.
  
Back to the real world scenario. Our client-side router is debouncing all the navigation handlers by default. I cannot really remember why it's done this way, but I guess it was a work around for some recursive update issues we had at some point, and, as you have noticed, this is an important detail required to provoke the bug. The way I actually encountered it was simply during one of my testing sessions when I, for some reason, decided to type in a route manually. Frankly, I don't believe it's too likely to happen the same way for a normal user, but who knows, there might be some other, easier ways to trigger the same behavior. 

Well, anyway, thanks for suggesting blurring of active element in the popstate handler, I'm going to add it to our router, so that we are safe, even if this is not going to be fixed.

**Comment by WebReflection** [comment]
I will try to fix it, but I've suggested the workaround because I don't know yet if I can.
Thanks for explaining the reason to denounce, it makes sense

**Comment by WebReflection** [comment]
OK, as I cannot possibly do anything on blur events since the node results still live when blur happens, I think you need to re-consider the debouncing of the update, since updating it right away does not produce errors, as well as keeping the state consistent, as well as using `setTimeout(update, 0, true);` instead of `debounce`.

This seems to me like a very debouncing specific issue, and if I solve it through any hack, the result is that the `Mojn` is never shown on the document, so the logic would be also broken.

In few words, this might be a won't fix that is worth a bug in Chrome.
